### PR TITLE
Reworked codec pipelines

### PR DESCRIPTION
**Pull request by normanrz**
This PR refactors the codec pipelines in the v3 codebase. There are now a new default implementation: 
- `BatchedCodecPipeline`, which divides the chunk batches into configurable "mini-batches". In a mini-batch all steps are run in lock step (e.g. fetching from store, decoding, encoding, writing to store). Multiple mini-batches are processed concurrently.

This PR moves a lot of code from the Array to the codec pipeline, which is an opportunity to share more code between the `Array` and the `ShardingCodec`. To make that work new `ByteGetter` and `ByteSetter` protocols to generalize the existing `StorePath` are introduced.

It also changes the Codec API by making `decode` and `encode` methods take chunk batches.

TODO:
* [ ] Add unit tests and/or doctests in docstrings
* [ ] Add docstrings and API docs for any new/modified user-facing classes and functions
* [ ] New/modified features documented in docs/tutorial.rst
* [ ] Changes documented in docs/release.rst
* [ ] GitHub Actions have all passed
* [ ] Test coverage is 100% (Codecov passes)

### CODE REVIEW

File: src/zarr/v3/codecs/batched_pipeline.py
Comment (akshaysubr) [inline src/zarr/v3/codecs/batched_pipeline.py:223]: Does this not require a similar `None` guard like with `_read_key` above?

File: src/zarr/v3/codecs/batched_pipeline.py
Comment (akshaysubr) [inline src/zarr/v3/codecs/batched_pipeline.py:250]: Any way to parallelize these portions between reading and decoding and subsequent encoding?

File: src/zarr/v3/metadata.py
Comment (akshaysubr) [inline src/zarr/v3/metadata.py:118]: Does this mean the use the batched pipeline by default always? That might not be the worst thing, but wondering if it might be useful to implement some mechanism to pass a policy to `parse_codecs` that could allow implementing an `InterleavedCodecPipeline` or something down the line?

**Comment by akshaysubr** [comment]
@normanrz Overall, this looks quite good to me. A couple of questions I had:

1. It makes sense for the default behavior to be just dispatching decoding to multiple threads in `decode_batch` and similarly for encoding. For codecs that want a different parallelization strategy, the idea would be that they just override `decode_batch` right?
2. All of the array types are currently `np.ndarray`. Would it be possible to generalize this to something that is agnostic to the specific library and uses the buffer protocol, `__array_interface__`, `__cuda_array_interface__` or [DLPack](https://dmlc.github.io/dlpack/latest/python_spec.html)? This would allow using numpy arrays as usual, but also cupy arrays, PyTorch tensors, JAX arrays, etc.

File: src/zarr/v3/metadata.py
Comment (normanrz) [reply src/zarr/v3/metadata.py:118]: I definitely want to add a `InterleavedCodecPipeline` as well. We could make that configurable through the runtime configuration.

File: src/zarr/v3/codecs/batched_pipeline.py
Comment (normanrz) [reply src/zarr/v3/codecs/batched_pipeline.py:250]: We could certainly parallelize that with `concurrent_map`. Not sure how that would be parallelized using the GPU.

**Comment by normanrz** [comment]
I refactored the codec pipeline quite a bit in the last commit.
- There are now a `BatchedCodecPipeline` and a `InterleavedCodecPipeline` building upon an abstract `CodecPipeline`
- There are new protocol types for `ByteGetter` and `ByteSetter`. These are basically a generalization of `StorePath`. This allows the sharding codec to use the same codec pipeline abstraction as the array. That moves us one step closer to treating shards as sub-arrays.
- The `decode`, `decode_partial`, `encode`, `encode_partial` now have the batched interface. 
- This still uses `np.ndarray` everywhere. I leave that to @akshaysubr to change to a generalized array interface. 

Currently, the choice of codec pipeline is hard-coded. I am still looking for a way to specify that. Should that go into `RuntimeConfiguration`? @d-v-b @jhamman

File: src/zarr/v3/abc/codec.py
Comment (d-v-b) [inline src/zarr/v3/abc/codec.py:33]: I'm having trouble understanding why this function exists just from its signature -- could you add a docstring that gives some context for it?

Comment (d-v-b) [reply src/zarr/v3/abc/codec.py:33]: I think I'm also struggling to understand the role of `None` in the chunk encoding / decoding process

Comment (normanrz) [reply src/zarr/v3/abc/codec.py:33]: Since this is the batch mode, we're dealing with multiple chunk requests. If the store returns `None` (i.e. missing chunk = fill value-only chunk) for any of these, we don't try to decode these. The `noop_for_none` function is just a higher order function so that the decode (or encode) functions don't have to do the early return on their own.

Comment (d-v-b) [reply src/zarr/v3/abc/codec.py:33]: Do you think we could simplify the logic here if stores didn't return `None`?

Comment (normanrz) [reply src/zarr/v3/abc/codec.py:33]: Not really. In that case, the codec pipeline would have to handle KeyError or similar.

Comment (d-v-b) [reply src/zarr/v3/abc/codec.py:33]: > Not really. In that case, the codec pipeline would have to handle KeyError or similar.

But if handling the key error was in just one place, and after handling keyerror the empty keys were turned into arrays of `fill_value` (or just ignored completely, if there was already a buffer pre-filled with `fill_value`), then I think there would be no need to handle `None` in any of the chunk processing logic, which would be kind of nice

Comment (normanrz) [reply src/zarr/v3/abc/codec.py:33]: I might be wrong but handling a `KeyError` or a `None` is the same complexity in the chunk processing logic. Both still have to happen per-chunk and coordinated with the other chunks in a batch request.

Comment (d-v-b) [reply src/zarr/v3/abc/codec.py:33]: I wonder if missing chunks should never enter the encoding / decoding process at all?

if we handle missing chunks early (e.g., right when we fetch them from the store, for decoding, or when we fetch them from the array, for encoding), then the chunk encoding / decoding routines never need to handle missing chunks, which means that `noop_for_none` wouldn't need to exist, because `chunk` could never be `None`, and likewise for all the other methods that take `Optional[np.ndarray]` as an argument.

Comment (normanrz) [reply src/zarr/v3/abc/codec.py:33]: The codec pipeline does the fetching so that code needs to live there. It also merges chunks into the decoded ndarray, for that it also needs to know which chunks are missing to fill them with fill_value. Using None to propagate that information seems ok.

**Comment by normanrz** [comment]
I think this PR is ready for a final review. I updated the PR description with the major changes. The only thing missing from my pov is the user-configurable batch size and codecpipeline selection. I'll add that after #1855 lands.

File: src/zarr/abc/codec.py
Comment (d-v-b) [inline src/zarr/abc/codec.py:33]:
```suggestion
    func: Callable[[T, ArraySpec, RuntimeConfiguration], Awaitable[U | None]],
```

File: src/zarr/abc/codec.py
Comment (d-v-b) [inline src/zarr/abc/codec.py:42]:
```suggestion
) -> Callable[[T | None, ArraySpec, RuntimeConfiguration], Awaitable[U | None]]:
    async def wrap(
        chunk: T | None, chunk_spec: ArraySpec, runtime_configuration: RuntimeConfiguration
    ) -> U | None:
        if chunk is None:
            return None
        return await func(chunk, chunk_spec, runtime_configuration)

    return wrap
```

File: src/zarr/abc/codec.py
Comment (normanrz) [reply src/zarr/abc/codec.py:33]: Can we do these changes automated?

File: src/zarr/abc/codec.py
Comment (d-v-b) [inline src/zarr/abc/codec.py:61]:
```suggestion
@runtime_checkable
class ByteGetter(Protocol):
    async def get(
        self, byte_range: tuple[int, int | None] | None = None
    ) -> BytesLike | None: ...


@runtime_checkable
class ByteSetter(Protocol):
    async def get(
        self, byte_range: tuple[int, int | None] | None = None
    ) -> BytesLike | None: ...

    async def set(self, value: BytesLike, byte_range: tuple[int, int] | None = None) -> None: ...

    async def delete(self) -> None: ...

```

File: src/zarr/abc/codec.py
Comment (d-v-b) [inline src/zarr/abc/codec.py:94]:
```suggestion
        chunk_arrays_and_specs: Iterable[tuple[np.ndarray | None, ArraySpec]],
        runtime_configuration: RuntimeConfiguration,
    ) -> Iterable[np.ndarray | None]:
```

File: src/zarr/abc/codec.py
Comment (d-v-b) [inline src/zarr/abc/codec.py:117]:
```suggestion
        chunk_arrays_and_specs: Iterable[tuple[np.ndarray | None, ArraySpec]],
        runtime_configuration: RuntimeConfiguration,
    ) -> Iterable[np.ndarray | None]:
```

File: src/zarr/abc/codec.py
Comment (d-v-b) [inline src/zarr/abc/codec.py:142]:
```suggestion
    async def decode_batch(
        self,
        chunk_bytes_and_specs: Iterable[tuple[BytesLike | None, ArraySpec]],
        runtime_configuration: RuntimeConfiguration,
    ) -> Iterable[np.ndarray | None]:
```

File: src/zarr/abc/codec.py
Comment (d-v-b) [inline src/zarr/abc/codec.py:165]:
```suggestion
    async def encode_batch(
        self,
        chunk_arrays_and_specs: Iterable[tuple[np.ndarray | None, ArraySpec]],
        runtime_configuration: RuntimeConfiguration,
    ) -> Iterable[BytesLike | None]:
```

File: src/zarr/abc/codec.py
Comment (d-v-b) [inline src/zarr/abc/codec.py:191]:
```suggestion
    async def decode_partial_batch(
        self,
        batch_info: Iterable[tuple[ByteGetter, SliceSelection, ArraySpec]],
        runtime_configuration: RuntimeConfiguration,
    ) -> Iterable[np.ndarray | None]:
```

File: src/zarr/abc/codec.py
Comment (d-v-b) [inline src/zarr/abc/codec.py:216]:
```suggestion
        batch_info: Iterable[tuple[ByteSetter, np.ndarray, SliceSelection, ArraySpec]],
```

File: src/zarr/abc/codec.py
Comment (d-v-b) [inline src/zarr/abc/codec.py:243]:
```suggestion
        chunk_bytes_and_specs: Iterable[tuple[BytesLike | None, ArraySpec]],
        runtime_configuration: RuntimeConfiguration,
    ) -> Iterable[BytesLike | None]:
```

File: src/zarr/abc/codec.py
Comment (d-v-b) [inline src/zarr/abc/codec.py:266]:
```suggestion
        chunk_bytes_and_specs: Iterable[tuple[BytesLike | None, ArraySpec]],
        runtime_configuration: RuntimeConfiguration,
    ) -> Iterable[BytesLike | None]:
```

File: src/zarr/codecs/pipeline/batched.py
Comment (d-v-b) [inline src/zarr/codecs/pipeline/batched.py:33]:
```suggestion
def _unzip2(iterable: Iterable[tuple[T, U]]) -> tuple[list[T], list[U]]:
    out0: list[T] = []
    out1: list[U] = []
    for item0, item1 in iterable:
```

File: src/zarr/abc/codec.py
Comment (normanrz) [reply src/zarr/abc/codec.py:33]: pylint has https://pylint.readthedocs.io/en/stable/user_guide/messages/refactor/consider-alternative-union-syntax.html

File: src/zarr/codecs/pipeline/batched.py
Comment (d-v-b) [inline src/zarr/codecs/pipeline/batched.py:30]: what's the difference between this function and zip with tuple unpacking (the last line in my example)?

```python
>>> a = ('a','b','c')
>>> b = (1, 2, 3)
>>> tuple(zip(a,b))
(('a', 1), ('b', 2), ('c', 3))
>>> z = tuple(zip(a,b))
>>> tuple(zip(*z))
(('a', 'b', 'c'), (1, 2, 3))
```

File: src/zarr/abc/codec.py
Comment (d-v-b) [reply src/zarr/abc/codec.py:33]: i'm not aware of an automated way to do it, but it would be a lot nicer than the manual way. you can also do an IDE-driven style where you remove the `Union` import from typing and find all the errors that result.

File: src/zarr/codecs/pipeline/hybrid.py
Comment (d-v-b) [inline src/zarr/codecs/pipeline/hybrid.py:28]: can we use [`itertools.batched`](https://docs.python.org/3/library/itertools.html#itertools.batched) instead?

Comment (normanrz) [reply src/zarr/codecs/pipeline/hybrid.py:28]: Unfortunately, no. `itertools.batched` is 3.12+

Comment (d-v-b) [reply src/zarr/codecs/pipeline/hybrid.py:28]: :(

File: src/zarr/codecs/pipeline/batched.py
Comment (normanrz) [reply src/zarr/codecs/pipeline/batched.py:30]: Functionally equivalent, but mypy didn't understand the types. I got `Tuple[List[T | U], ...]` as return type.

File: src/zarr/abc/codec.py
Comment (d-v-b) [inline src/zarr/abc/codec.py:94]: What would be the problem with making `decode` just a special case of `decode_batch`, where the length of `chunk_arrays_and_specs` is 1? 

Also, it seems odd that `decode_batch` with a batch size of 1 could be nullable, while `decode` with the same argument would not be.

Comment (normanrz) [reply src/zarr/abc/codec.py:94]: That would be possible. If we want to make it easier for the community to develop codecs, the batch-style is a bit more complicated. We can provide other means of making that easier, though (helper functions, or `Base*Codec` classes).

Comment (d-v-b) [reply src/zarr/abc/codec.py:94]: to be clear, even if there are performance differences between "do 1" and "do N", I think we should consider hiding that behind a single "do N" api that dispatches to the efficient sub-case as needed.

File: src/zarr/codecs/gzip.py
Comment (normanrz) [reply src/zarr/codecs/gzip.py:47]: That is what the `BytesBytesBatchMixin` does.
Btw, let me know if you know of a better name than `BytesBytesBatchMixin`. Fwiw, they are not real mixins more like base classes.

File: src/zarr/abc/codec.py
Comment (normanrz) [reply src/zarr/abc/codec.py]: Done

File: src/zarr/codecs/pipeline/batched.py
Comment (d-v-b) [inline src/zarr/codecs/pipeline/batched.py:44]: Since `CodecPipeline` defines `read_batch` and `write_batch`, what is the relationship between those methods and the new functionality offered by this class? I think a clarifying docstring for the class might be useful here, because I don't find the inheritance relationship intuitive (and ditto for the `HybridPipeline` -- I have no intuition for that that one is for :) )

File: src/zarr/codecs/bytes.py
Comment (d-v-b) [inline src/zarr/codecs/bytes.py:28]: Will `ArrayBytesCodec` be used anywhere other than as the base class fro `ArrayBytesCodecBatchMixin`? If not, we might want to consider simplifying the inheritance structure a bit.

File: src/zarr/codecs/mixins.py
Comment (d-v-b) [inline src/zarr/codecs/mixins.py:105]: The amount of logic duplicated here really stands out to me. I wonder if there is an abstraction we are missing that could simplify this

File: src/zarr/codecs/bytes.py
Comment (normanrz) [reply src/zarr/codecs/bytes.py:28]: I would anticipate that folks want to build a codec that implements their own batching, e.g. in rust or on the GPU. That is why, we should keep both classes.

File: src/zarr/codecs/pipeline/hybrid.py
Comment (d-v-b) [inline src/zarr/codecs/pipeline/hybrid.py:35]: What are the operational differences between this class and the `BatchedCodecPipeline`?

File: src/zarr/codecs/pipeline/hybrid.py
Comment (d-v-b) [inline src/zarr/codecs/pipeline/hybrid.py:37]: Having a sibling class as an attribute is rather surprising. I see that this attribute is used quite extensively in this class, in combination with the `batched` function. This makes me wonder whether we need `HybridCodecPipeline` at all, as opposed to just putting a `batch_size` parameter on `BatchedCodecPipeline`

File: src/zarr/codecs/pipeline/hybrid.py
Comment (normanrz) [reply src/zarr/codecs/pipeline/hybrid.py:35]: This is a good question. `BatchedCodecPipeline` executes all steps in lock step for all chunks that are requested. `HybridCodecPipeline` adds a batch size where each batch can run asynchronously interleaved. We could consolidate both into 1 class. That would leave the abstract `CodecPipeline` (for implementing custom codec pipelines) and one implementation.

File: src/zarr/codecs/mixins.py
Comment (normanrz) [reply src/zarr/codecs/mixins.py:105]: Can you elaborate? The logic is reused through the `batching_helper`. Large parts of the signature are duplicated but differ in details due to the nature of the 3 separate codec classes. What abstraction would you propose?

**Comment by madsbk** [comment]
I think it would be helpful with some docstrings. At least, I find it hard to follow the intention without any help :)

File: src/zarr/codecs/pipeline/hybrid.py
Comment (normanrz) [reply src/zarr/codecs/pipeline/hybrid.py]: I got rid of the `HybridCodecPipeline` and added a `batch_size` parameter to the `BatchedCodecPipeline`.

File: src/zarr/codecs/mixins.py
Comment (normanrz) [reply src/zarr/codecs/mixins.py]: I was able to get rid of a lot of this code by using `typing.Generic`.

**Comment by normanrz** [comment]
I added a few doc strings and implemented the abstract Codec classes with Generics.

File: src/zarr/codecs/batched_codec_pipeline.py
Comment (jhamman) [inline src/zarr/codecs/batched_codec_pipeline.py:62]: this can come later but we're going to want some additional documentation on the behavior here. Reading this, I'm not entirely sure I get it.

File: src/zarr/codecs/batched_codec_pipeline.py
Comment (jhamman) [inline src/zarr/codecs/batched_codec_pipeline.py:130]:
```suggestion
                + "writes, which may lead to inefficient performance.",
                stacklevel=3
            )
```

File: src/zarr/codecs/batched_codec_pipeline.py
Comment (jhamman) [inline src/zarr/codecs/batched_codec_pipeline.py:153]: let's add a docstring or comment here. The comment should answer why having an array_array_codec or bytes_bytes_codec in the pipeline means `supports_partial_decode is False`.

File: src/zarr/codecs/batched_codec_pipeline.py
Comment (jhamman) [inline src/zarr/codecs/batched_codec_pipeline.py:156]: same comment as above.

**Review by jhamman — APPROVED** [review]
@normanrz -- this is an impressive piece of work. I should admit that the size of it made it hard to review so I just have a few comments. Overall, I think its a big step forward and I want to get my hands on it so I favor moving it into the v3 branch asap.

**Review by d-v-b — APPROVED** [review]

File: src/zarr/codecs/mixins.py
Comment (d-v-b) [reply src/zarr/codecs/mixins.py:25]: This doesn't need to be addressed here, so feel free to ignore for now

Comment (normanrz) [reply src/zarr/codecs/mixins.py:25]: How would you feel about this?
```python
class _Codec(Generic[CodecInput, CodecOutput], Metadata):
    ...

    async def _decode_single(self, chunk_data: CodecOutput, chunk_spec: ArraySpec) -> CodecInput:
        raise NotImplementedError

    async def decode(
        self, chunk_data_and_specs: Iterable[tuple[CodecOutput | None, ArraySpec]]
    ) -> Iterable[CodecInput | None]:
        return await batching_helper(self.decode_single, chunk_data_and_specs)

    # same for encode
    ...
```
Batch-aware codecs would then only override the `decode` method and ignore the `_decode_single` method. That would be fine because `_decode_single` is a protected method not intended to be part of the public interface. Single-chunk codecs could override `_decode_single` and won't have to care about the batching. We could drop the batch mixins, then.

File: src/zarr/codecs/_v2.py
Comment (normanrz) [inline src/zarr/codecs/_v2.py:104]: @madsbk I added a very naive Buffer integration in these codecs. Could you please help me get that right?

Comment (normanrz) [reply src/zarr/codecs/_v2.py:104]: I'll merge this PR for now. We can work on a new PR to optimize this.

File: src/zarr/abc/codec.py
Comment (normanrz) [reply src/zarr/abc/codec.py:61]: What about naming the function `evolve_from_array_spec`?

File: src/zarr/codecs/mixins.py
Comment (d-v-b) [reply src/zarr/codecs/mixins.py:25]: sounds good!

File: src/zarr/abc/codec.py
Comment (d-v-b) [reply src/zarr/abc/codec.py:61]: i think that works

Comment (d-v-b) [reply src/zarr/abc/codec.py:61]: alternatively, we could make `from_array_spec` a class method that takes keyword arguments to cover the attributes that the input array spec doesn't convey

File: src/zarr/codecs/_v2.py
Comment (madsbk) [reply src/zarr/codecs/_v2.py:104]: Sounds good!

**Pull request merged**
